src/c-tips/c-tips.md

```diff
@@ -1,3 +1,209 @@
+
+[macro system]: https://doc.rust-lang.org/book/second-edition/appendix-04-macros.html
+
+In general, where you might have used a C preprocessor macro, you probably want
+to see if a macro-by-example can do the job instead. They can be defined in
+your crate and easily used by your own crate or exported for other users. Be
+aware that since they must expand to complete expressions, statements, items,
+or patterns, some use cases of C preprocessor macros will not work, for example
+a macro that expands to part of a variable name or an incomplete set of items
+in a list.
+
+As with Cargo features, it is worth considering if you even need the macro. In
+many cases a regular function is easier to understand and will be inlined to
+the same code as a macro. The `#[inline]` and `#[inline(always)]` [attributes]
+give you further control over this process, although care should be taken here
+as well — the compiler will automatically inline functions from the same crate
+where appropriate, so forcing it to do so inappropriately might actually lead
+to decreased performance.
+
+[attributes]: https://doc.rust-lang.org/reference/attributes.html#inline-attribute
+
+Explaining the entire Rust macro system is out of scope for this tips page, so
+you are encouraged to consult the Rust documentation for full details.
+
+## Build System
+
+Most Rust crates are built using Cargo (although it is not required). This
+takes care of many difficult problems with traditional build systems. However,
+you may wish to customise the build process. Cargo provides [`build.rs`
+scripts] for this purpose. They are Rust scripts which can interact with the
+Cargo build system as required.
+
+[`build.rs` scripts]: https://doc.rust-lang.org/cargo/reference/build-scripts.html
+
+Common use cases for build scripts include:
+
+* provide build-time information, for example statically embedding the build
+  date or Git commit hash into your executable
+* generate linker scripts at build time depending on selected features or other
+  logic
+* change the Cargo build configuration
+* add extra static libraries to link against
+
+### Cross-Compiling
+
+Using Cargo for your build system also simplifies cross-compiling. In most
+cases it suffices to tell Cargo `--target thumbv6m-none-eabi` and find a
+suitable executable in `target/thumbv6m-none-eabi/debug/myapp`.
+
+For platforms not natively supported by Rust, you will need to build `libcore`
+for that target yourself. On such platforms, [Xargo] can be used as a stand-in
+for Cargo which automatically builds `libcore` for you.
+
+[Xargo]: https://github.com/japaric/xargo
+
+## Iterators vs Array Access
+
+In C you are probably used to accessing arrays directly by their index:
+
+```c
+int16_t arr[16];
+int i;
+for(i=0; i<sizeof(arr)/sizeof(arr[0]); i++) {
+    process(arr[i]);
+}
+```
+
+In Rust this is an anti-pattern: indexed access can be slower (as it needs to
+be bounds checked) and may prevent various compiler optimisations.
+
```

Comment:
Something like this addition perhaps, to really highlight this very important distinction? Edit as seen fit. :)

```suggestion
It is important to see this distinction in Rust vs C, as Rust needs to check for out-of-bounds access when manually indexing an array to guarantee memory safety, while C will happily index outside the array if given an index that will do so.  
```